### DISCUSSION TITLE
Board a spacecraft and flyby planets, moons and a comet yourself

### DISCUSSION THREAD

**Discussion by stephanpls** [Stellarium/stellarium#2178] [Show and tell]
To make Stellarium even more dynamic, you can experience now - 'onboard' a spacecraft - flybys of planets, moons and a comet that really have taken place.

**Installation**
1. Unzip this attachment:
[Stellarium_flybys.zip](https://github.com/Stellarium/stellarium/files/7876946/Stellarium_flybys.zip)
2. Put the .ssc files in the Stellarium installation directory: ..\Stellarium\scripts\ (for instance C:\Program Files (x86)\Stellarium\scripts\). You may encounter a message about administrator rights, so give yourself these rights then
3. Put file ssystem_spacecraft_and_flybys.ini (with orbit data of the spacecraft and planet, moon etc.) anywhere you like, but best in the installation directory (for instance C:\Program Files (x86)\Stellarium\data\)
4. Start Stellarium
5. Run the Solar System Editor (SSE) plugin via F2, tab Plug-ins, scroll at the left and select the SSE
6. In the SSE: click Configure, then the last option ' Import and add Solar System minor objects from file...'
7. Select ssystem_spacecraft_and_flybys.ini
8. The orbital elements should have been loaded now

**Running a script**
1. Press F2, tab Scripts, scroll at the left where all just installed scripts start with 'Flyby..'
2. Select the script
3. You are now 'onboard' the spacecraft and watch the flyby unfold. You see with your naked eye (FOV 27°) the object to flyby in the middle of the screen. In the beginning it seems slow, since the script starts three days before the flyby. But, the time rate has been set to 2000 so it takes about two minutes to flyby. You better decrease the time rate to better study the actual flyby!

I hope that all is working on your system. Have fun!

**Notes**
1. ssystem_spacecraft_and_flybys.ini contains the orbital elements of the involved spacecraft and planet etc., and ARE ONLY VALID for the moment of the actual flyby. The bigger the planet and the closer the flyby, the stronger the gravity influence on the spacecraft orbit. So, the longer before and after the flyby you are, the less valid the elements are. That's to say when the script starts the spacecraft position is not exactly correct. But that does not matter that much for the experience you get, which is quite real and dynamic
2. I added elements of comet 19P/Borrelly because the current elements of it (with orbit_good=285 days only) in Stellarium give positions far off at the time of the flyby
3. Flyby_1968-12-24_Apollo8_S-IVB@Moon shows the Moon flyby of the third stage of the Saturn V of the Apollo 8. Note that this is not the Apollo CSM, with astronauts Anders, Borman and Lovell. They were the first humans to leave the gravitational influence of the Earth and travel to the Moon, on Christmas Eve 1968. The S-IVB passes the Moon above Mare Crisium (at the "right"). The CSM Moon orbit insertion was at the other side of the Moon ("left"), but anyway, the distance between CSM and S-IVB at the flyby was not more than a few thousand km. So what you see in the animation is a quite good impression of what the astronauts saw. Unfortunately the orbital elements of the CSM are not available at https://ssd.jpl.nasa.gov/horizons/
4. Using script Flyby_2004-08-20_MarsExpress@Phobos, about 1450 km away from the Mars Express Phobos suddenly disappears in an odd way. At least it happens with the Stellarium version I use, 21.2. See picture, Phobos becomes a ring! You may see it if you decrease the time rate considerably. It seems to be a bug (?)
![image](https://user-images.githubusercontent.com/96252651/149660301-9fe41367-41ed-4d23-8280-28ac9b59a979.png)

**Comment by gzotti** [comment]
Thanks, that's a sort of preview of a future "interplanetary spacecraft" plugin. 

The ring looks like a near clipping plane issue when you come too close to the object. 

@alex-w can we host this among the scripts on stellarium.org, or should we even include it in the distribution, with description in chapter 16?

**Reply by alex-w** [reply]
I think we may add 2-3 wonderful fly-by's into main package and share put all collection on the stellarium.org

@stephanpls is it acceptable for you?

**Comment by stephanpls** [comment]
Yes. And I could add more interesting flybys if requested by the community, since there are many and I did not study all of them of being 'beautiful'. It's no more a laborious issue for me. I automated the proces of getting orbital elements, processing them, and writing the .ini and the .ssc.  I just need a spacecraft name, a flyby object and the date/time. Then click and ready.

**Reply by gzotti** [reply]
Well then please add: Luna 3, Vega 1/Halley, Vega 2/Halley, Giotto/Halley. Note that Halley is available even as 3D model. But you may need to get orbital elements for Halley in those weeks as well.  I cannot say how to add Rosetta to Churyumov-Gerasimenko as this was no flyby, not sure if there are piecewise working heliocentric orbital elements available. 

Maybe add an orbit_good value to the elements to let users estimate if the elements are reasonable at some particular date.

**Comment by alex-w** [comment]
@stephanpls any news?

**Comment by stephanpls** [comment]
Heavily occupied with other non-Stellarium things .... Maybe this weekend I'll have a look at it.

**Reply by stephanpls** [reply]
Of course I first tried to show a Halley flyby, but unfortunately https://ssd.jpl.nasa.gov/horizons/time_spans.html (click 'Spacecraft' on the left to get all spacecraft available) does not provide orbital elements for Vega 1, Vega 2 and Giotto.
Nor does it have elements for the Luna 3, the first spacecraft that photographed the other side of the Moon.

**Rosetta@P67/Churyumov-Gerasimenko**
Please find attached a zip with the orbital data and Stellarium script. Follow the indications in the first post of this thread to get it running in Stellarium.
True, this was not a flyby; Rosetta put itself in an orbit around the comet and even got its small lander on the comet surface. IMHO this is not the most beautiful of all flybys uploaded in the first post, but anyway it gives an impression.  

[Rosetta@67P_Churyumov_Gerasimenko.zip](https://github.com/Stellarium/stellarium/files/8102071/Rosetta%4067P_Churyumov_Gerasimenko.zip)

**Reply by alex-w** [reply]
Please add into archive README file with instructions for users and please add header part into script

**Comment by isenberg** [comment]
Nice! I'll try that out with adding SPICE data for Mars orbiters, see #3058 .

**Comment by hn-88** [comment]
Thank you, @stephanpls for creating these scripts - I've extended the Saturn flyby script by adding screen capture, and have created a couple of Saturn flyby videos, in fulldome format suitable for planetariums,
https://hnsws.blogspot.com/2024/02/creating-stellarium-flybys.html

The modified script is at
https://github.com/hn-88/stellarium-scripts/blob/main/Flyby_1981-08-25-4pm_Voyager2%40Saturn%20-%20Vid.ssc
and direct links to the youtube videos are 

https://www.youtube.com/watch?v=eBu-8wRxs-w

https://www.youtube.com/watch?v=rpi5aDjuqwk
